opentreemap/treemap/udf.py

```diff
@@ -766,111 +877,115 @@ def datatype_by_field(self):
             try:
                 return float(value)
             except ValueError:
-                raise ValidationError(_('%(fieldname)s '
-                                        'must be a real number') %
-                                      {'fieldname': self.name})
+                raise complaint('%(fieldname)s must be a real number',
+                                fieldname=self.name)
         elif datatype == 'int':
             try:
                 if float(value) != int(value):
                     raise ValueError
 
                 return int(value)
             except ValueError:
-                raise ValidationError(_('%(fieldname)s must be an integer') %
-                                      {'fieldname': self.name})
-        elif datatype == 'user':
-            if isinstance(value, User):
-                return value
-
-            try:
-                pk = int(value)
-            except ValueError:
-                raise ValidationError(_('%(fieldname)s must be an integer') %
-                                      {'fieldname': self.name})
-            try:
-                return User.objects.get(pk=pk)
-            except User.DoesNotExist:
-                raise ValidationError(_('%(fieldname)s user not found') %
-                                      {'fieldname': self.name})
-
+                raise complaint('%(fieldname)s must be an integer',
+                                fieldname=self.name)
         elif datatype == 'date':
             if isinstance(value, (date, datetime)):
                 return value
 
             try:
                 valid_date = parse_date_string_with_or_without_time(value)
             except ValueError:
-                raise ValidationError(_('%(fieldname)s must be formatted as '
-                                        'YYYY-MM-DD') %
-                                      {'fieldname': self.name})
+                raise complaint(('%(fieldname)s must be formatted as '
+                                 'YYYY-MM-DD'),
+                                fieldname=self.name)
 
             # Ensure date UDF values contain a year >= 1900 so that
             # date formatting with `strftime` will work correctly.
             if valid_date.year < 1900:
-                raise ValidationError(_('%(fieldname)s year must be >= 1900')
-                                      % {'fieldname': self.name})
+                raise complaint('%(fieldname)s year must be >= 1900',
+                                fieldname=self.name)
 
             return valid_date
 
         elif 'choices' in datatype_dict:
             def _validate(val):
                 if val not in datatype_dict['choices']:
-                    raise ValidationError(
-                        _('Invalid choice (%(given)s). Expecting %(allowed)s')
-                        % {'given': val,
-                           'allowed': ', '.join(datatype_dict['choices'])})
+                    raise complaint(
```

Comment:
I am running into this validation error when trying to clear the value of a choice field.

- Create a UDF choice field on Tree
- Add a tree and set a UDF choice value.
- Refresh the tree detail page.
- Edit and try to set the choice field to an empty value.

<img width="521" alt="screen shot 2017-05-15 at 1 10 32 pm" src="https://cloud.githubusercontent.com/assets/17363/26077148/eb058db8-396f-11e7-83c1-41efa85ee674.png">

In a related test, I edited the value of a custom text field on a `treefolksaustin` tree that did not have a choice value selected and got an "invalid choice (None)" message. When I tried to refresh the map feature page it now crashes with a ValidationError

```
Traceback:
File "/usr/local/lib/python2.7/dist-packages/django/core/handlers/base.py" in get_response
  132.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/opt/app/core/opentreemap/treemap/decorators.py" in wrapper
  46.             return view_fn(request, instance, *args, **kwargs)
File "/usr/local/lib/python2.7/dist-packages/django_tinsel/decorators.py" in routed
  33.             return req_method(request, *args2, **kwargs2)
File "/opt/app/core/opentreemap/treemap/views/map_feature.py" in render_map_feature_detail
  159.                               **kwargs)
File "/opt/app/core/opentreemap/treemap/views/map_feature.py" in map_feature_detail
  86.         request, instance, feature_id, edit)
File "/opt/app/core/opentreemap/treemap/views/map_feature.py" in _map_feature_detail_context
  103.     context = ctx_fn(request, feature, edit=edit)
File "/opt/app/core/opentreemap/treemap/lib/map_feature.py" in context_dict_for_plot
  185.     context = context_dict_for_map_feature(request, plot, **kwargs)
File "/opt/app/core/opentreemap/treemap/lib/map_feature.py" in context_dict_for_map_feature
  407.         'title': feature.title(),
File "/opt/app/core/opentreemap/treemap/models.py" in title
  799.             tree = feature.current_tree()
File "/opt/app/core/opentreemap/treemap/models.py" in current_tree
  1001.         trees = list(self.tree_set.all())
File "/usr/local/lib/python2.7/dist-packages/django/db/models/query.py" in __iter__
  162.         self._fetch_all()
File "/usr/local/lib/python2.7/dist-packages/django/db/models/query.py" in _fetch_all
  965.             self._result_cache = list(self.iterator())
File "/usr/local/lib/python2.7/dist-packages/django/db/models/query.py" in iterator
  255.             obj = model_cls.from_db(db, init_list, row[model_fields_start:model_fields_end])
File "/usr/local/lib/python2.7/dist-packages/django/db/models/base.py" in from_db
  489.             new = cls(*values)
File "/opt/app/core/opentreemap/treemap/models.py" in __init__
  1104.         super(Tree, self).__init__(*args, **kwargs)
File "/opt/app/core/opentreemap/treemap/units.py" in __init__
  23.         super(Convertible, self).__init__(*args, **kwargs)
File "/opt/app/core/opentreemap/treemap/udf.py" in __init__
  1317.         self._setup_udfs(udfs_kwarg, hstore_udfs_kwarg)
File "/opt/app/core/opentreemap/treemap/udf.py" in _setup_udfs
  1349.                                        **hstore_udfs_kwarg)
File "/opt/app/core/opentreemap/treemap/udf.py" in __init__
  1160.                     self[k] = v
File "/opt/app/core/opentreemap/treemap/udf.py" in __setitem__
  1241.                 cleaned_value = udf.clean_value(val, datatype_dict=dtd)
File "/opt/app/core/opentreemap/treemap/udf.py" in clean_value
  977.                 _validate(value)
File "/opt/app/core/opentreemap/treemap/udf.py" in _validate
  974.                         allowed=', '.join(datatype_dict['choices']))

Exception Type: ValidationError at /treefolksaustin/features/2924118/
Exception Value: {u'tree.udf:Historic or champion tree? (tell us about it in comments!)': [u'Invalid choice (None). Expecting yes, no']}
```